apps/rand.c

```diff
@@ -95,8 +97,81 @@ int rand_main(int argc, char **argv)
     argc = opt_num_rest();
     argv = opt_rest();
     if (argc == 1) {
-        if (!opt_long(argv[0], &num) || num <= 0)
+        int factoridx = 0;
+        int shift = 0;
+
+        /*
+         * special case for requesting the max allowed
+         * number of random bytes to be generated
+         */
+        if (!strcmp(argv[0], "max")) {
+            scaled_num = SIZE_MAX;
+        } else {
+            /*
+             * iterate over the value and check to see if there are
+             * any non-numerical chars
+             * A non digit suffix indicates we need to shift the
+             * number of requested bytes by a factor of:
+             * k = 1024^1 (1 << (10 * 1))
+             * m = 1024^2 (1 << (10 * 2))
+             * g = 1024^3 (1 << (10 * 3))
+             * t = 1024^4 (1 << (10 * 4))
+             * which can be achieved by bit-shifting the number
+             */
+            while (argv[0][factoridx]) {
+                if (!isdigit((int)(argv[0][factoridx]))) {
+                    switch(argv[0][factoridx]) {
+                    case 'k':
+                    case 'K':
+                        shift = 10;
+                        break;
+                    case 'm':
+                    case 'M':
+                        shift = 20;
+                        break;
+                    case 'g':
+                    case 'G':
+                        shift = 30;
+                        break;
+                    case 't':
+                    case 'T':
+                        shift = 40;
+                        break;
+                    default:
+                        BIO_printf(bio_err, "Invalid size suffix %s\n",
+                                   &argv[0][factoridx]);
+                        goto opthelp;
+                    }
+                    break;
+                }
+                factoridx++;
+            }
+
+            if (shift != 0 && strlen(&argv[0][factoridx]) != 1) {
+                BIO_printf(bio_err, "Invalid size suffix %s\n",
+                           &argv[0][factoridx]);
+                goto opthelp;
+            }
+        }
+        /* Remove the suffix from the arg so that opt_long works */
+        if (shift != 0)
+            argv[0][factoridx] = '\0';
+
+        if ((scaled_num != SIZE_MAX) && (!opt_long(argv[0], &num) || num <= 0))
```

Comment:
`opt_long()` calls `strtol()`, which deals with signed integers, so only up to `2147483647` on a 32-bit platform (and on Windows) (rather than 4294967295). `strtol()` also has a bunch of behaviours that we can't use, such as negative numbers and hex, and the annoying habit of treating leading `0` as introducing an octal number.

It might be an unusual suggestion, but would it be better to roll our own, given we've already gone through the input number a character at a time?

Also, since we now prevent hex input which was previously accepted, is there any generic documentation that says something like "numbers may be given in decimal, hexadecimal or octal format" that needs to be updated for this particular use?

Reply:
num is the prescaled value, if you look below, you'll see where it gets upscaled to the scaled_num variable which is size_t.

As for using strtol, thats the conversion mechanism that has always been there.  We can certainly change it, but that seems a bit out of scope for this PR

Reply:
> As for using strtol, thats the conversion mechanism that has always been there. We can certainly change it, but that seems a bit out of scope for this PR

I think it's fine as it is. Nobody will pass such ridiculously large numbers (which were not possible previously either) now that they can use the suffixes.

I'd expect `opt_long()` to return a value `num <= 0`, so you would end up with the usage message, right? (BTW: every `goto opthelp` should be preceded by an error message, also the one in line 161.)

Reply:
whats to be said on line 161?  opt_long already reports an unparseable value via opt_number_error